<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python

In [3]:
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port=5432
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [4]:
#n=10
query= f'''SELECT *
               FROM vacancies'''

In [5]:
# текст запроса
query_3_1 = f'''SELECT *
            FROM vacancies 
            
'''


In [6]:
df = pd.read_sql_query(query, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/137330334.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

In [7]:
# текст запроса
query = f'''SELECT count(distinct(vacancies)) 
            FROM vacancies
'''

In [8]:
# результат запроса
df = pd.read_sql_query(query, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/1115395102.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [9]:
# текст запроса
query3_2 = f'''SELECT count(id) 
            FROM employers
'''

In [10]:
# результат запроса
df = pd.read_sql_query(query3_2, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/3372558505.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_2, connection)


count
0  23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [11]:
# текст запроса
query3_3 = f'''SELECT count(id) 
            FROM areas
'''

In [12]:
# результат запроса
df = pd.read_sql_query(query3_3, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/1045105937.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_3, connection)


count
0   1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [13]:
# текст запроса
query3_4 = f'''SELECT count(id) 
            FROM industries
'''

In [14]:
# результат запроса
df = pd.read_sql_query(query3_4, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/3279677683.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query3_4, connection)


count
0    294

***

In [15]:
# выводы по предварительному анализу данных

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [16]:
# текст запроса
query4_1 = f'''SELECT a.name,COUNT(v.id)  
            FROM vacancies AS v 
            LEFT JOIN areas AS a ON v.area_id=a.id
            GROUP BY a.name
            ORDER BY COUNT(v.id) DESC
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query4_1, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/169690575.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_1, connection)


name  count
0             Москва   5333
1    Санкт-Петербург   2851
2              Минск   2112
3        Новосибирск   2006
4             Алматы   1892
..               ...    ...
764       Тарко-Сале      1
765    Новоаннинский      1
766            Бирск      1
767           Сасово      1
768             Анна      1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [18]:
# текст запроса
query4_2 = f'''SELECT count(id) 
            FROM vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query4_2, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/3983020382.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [20]:
# текст запроса
query4_3 = f'''SELECT ROUND(avg(salary_from),0) AS avg_from
                , round(avg(salary_to),0) AS avg_to

            FROM vacancies
            WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query4_3, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/570421473.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_3, connection)


avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [22]:
# текст запроса
query4_4 = f'''SELECT count(id), schedule, employment 
            FROM vacancies
            GROUP BY schedule, employment 
            ORDER BY count(id) DESC
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query4_4, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/1521009827.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_4, connection)


count          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [24]:
# текст запроса
query4_5 = f'''SELECT count(id),  experience
            FROM vacancies
            GROUP BY experience 
            ORDER BY count(id) 
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query4_5, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/2800791109.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query4_5, connection)


count          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

# выводы по детальному анализу вакансий
1. Наибольшее количество вакансий в  больших городах таких как : МоскваБ С-Петербург, Минск, Новосибирск, Алматы 
2. Менее чем в половине вакансий (24073 из 49197) заполнено хотя бы одно из двух полей с зарплатой
3. В среднем оплата труда варьируется от 71 тыс до 110,5 тыс
4. Работодатели предпочитают брать сотрудников **на полный день и на полную занятость**, вторым по популярности сочетанием является удаленная работа на полный день
5. Меньше всего работодателям требуются опытные сосискатели со стажем **более 6 лет**, а наиболее популярная категория сосикатели с начальным опытом от **1 года до 3**.

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [26]:
# текст запроса
query5_1 = f'''SELECT count(v.id), e.name
            FROM vacancies as v left join employers as e ON v.employer_id = e.id
            GROUP BY e.name
            ORDER BY count(v.id) DESC
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query5_1, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/3060966061.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_1, connection)


count                    name
0       1933                  Яндекс
1        491              Ростелеком
2        444                Тинькофф
3        428                    СБЕР
4        331           Газпром нефть
...      ...                     ...
14761      1            им. Тьюринга
14762      1        LLC Novalab Tech
14763      1   Компания Мир упаковки
14764      1  Лифинцев Олег Иванович
14765      1      ТрансЭнергоХолдинг

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [29]:
query5_2 = f'''
SELECT a.name AS region_name, 
       COUNT(v.employer_id) AS cnt_empl, 
       COUNT(v.id) AS cnt_vac
FROM vacancies AS v 
left JOIN areas AS a ON v.area_id = a.id
LEFT JOIN employers AS e ON a.id = e.area 

GROUP BY a.name
HAVING COUNT(v.id)=0
ORDER BY  cnt_empl DESC;
'''

In [30]:
# результат запроса
df = pd.read_sql_query(query5_2, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/1272371655.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_2, connection)


Empty DataFrame
Columns: [region_name, cnt_empl, cnt_vac]
Index: []

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [33]:
# текст запроса
query5_3 = f'''SELECT e.name AS empl_name, COUNT(distinct(v.area_id)) AS cnt_area 
            FROM vacancies as v 
            LEFT JOIN areas AS a on v.area_id=a.id
            LEFT JOIN employers AS e on v.employer_id=e.id
            GROUP BY empl_name
            ORDER BY cnt_area desc
'''

In [34]:
# результат запроса
df = pd.read_sql_query(query5_3, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/1671854320.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_3, connection)


empl_name  cnt_area
0                      Яндекс       181
1                  Ростелеком       152
2                  Спецремонт       116
3      Поляков Денис Иванович        88
4                   ООО ЕФИН         71
...                       ...       ...
14761                  UniSol         1
14762            UNISTORY LLC         1
14763                   UNIT6         1
14764     United Distribution         1
14765              UNITEL LLC         1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [35]:
# текст запроса
query5_4 = f'''SELECT COUNT(CASE WHEN ei.industry_id IS NULL THEN 1 END) AS cnt_ind
            FROM employers AS e 
            LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id
            
            
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query5_4, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/597290805.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_4, connection)


cnt_ind
0     8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [37]:
# текст запроса
query5_5 = f'''SELECT 
            e.name AS empl_name,
            COUNT(ei.industry_id) AS cnt_ind
            FROM employers AS e 
            LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id
            GROUP BY e.name
            HAVING COUNT(ei.industry_id) = 4
            ORDER BY empl_name
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query5_5, connection)
df.head(5)

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/1927874690.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_5, connection)


empl_name  cnt_ind
0  101 Интернет        4
1      21vek.by        4
2          2ГИС        4
3            2К        4
4   4 пикселя +        4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [39]:
# текст запроса
query5_6 = f'''SELECT COUNT(e.id) AS cnt_ind
            FROM employers AS e 
            LEFT JOIN employers_industries AS ei ON e.id = ei.employer_id
            LEFT JOIN industries AS i on i.id=ei.industry_id
            WHERE i.name='Разработка программного обеспечения'
            
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query5_6, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/2781422708.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_6, connection)


cnt_ind
0     3553

7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [41]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8#%D0%A1%D0%BF%D0%B8%D1%81%D0%BE%D0%BA_%D0%B3%D0%BE%D1%80%D0%BE%D0%B4%D0%BE%D0%B2-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D0%BE%D0%B2'
data = pd.read_html(url)
city_df = data[0]
city_list = list(city_df['Город'])
city_list

['Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Красноярск',
 'Нижний Новгород',
 'Челябинск',
 'Уфа',
 'Самара',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград']

In [42]:
# текст запроса
query5_7 = f'''with yandex_vacancies AS (SELECT (a.name ) AS city,
            count(v.id) AS cnt_vacancies
            FROM 
            vacancies AS v 
            LEFT JOIN employers AS e  ON e.id = v.employer_id
            LEFT JOIN areas AS a ON a.id=v.area_id
            WHERE e.name='Яндекс' and a.name IN ({', '.join(repr(city) for city in city_list)})
            GROUP BY a.name
            ORDER BY count(v.id))
           
            SELECT city, cnt_vacancies FROM yandex_vacancies
            UNION ALL
            SELECT 'Total', SUM(cnt_vacancies) FROM yandex_vacancies
            
'''

In [43]:
#результат запроса
df = pd.read_sql_query(query5_7, connection)

print(df)
print(f' количество строк {df.shape[0]}')


               city  cnt_vacancies
0              Омск           21.0
1         Челябинск           22.0
2        Красноярск           23.0
3         Волгоград           24.0
4             Пермь           25.0
5            Казань           25.0
6    Ростов-на-Дону           25.0
7               Уфа           26.0
8            Самара           26.0
9         Краснодар           30.0
10          Воронеж           32.0
11      Новосибирск           35.0
12  Нижний Новгород           36.0
13     Екатеринбург           39.0
14  Санкт-Петербург           42.0
15           Москва           54.0
16            Total          485.0
 количество строк 17


/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/3107441246.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_7, connection)


***

# выводы по анализу работодателей
- всего в базе представленна хотя бы одна вакансия у 14766 (ок 63%) работодателей из 23510
- в предоставленной базе имеется почти 15 тыс вакансий
- почти у трети работодателей (8419) не указана сфера деятельности
- среди тех компаний, у которых указана сфера деятельности, 3553 указали в качестве сферы деятельности: «Разработка программного обеспечения»  
- Наибольшее количество вакансий у компании Яндекс (1933) из них 485 в городах-миллионниках
- Яндекс представлен в наибольшем количестве регионов (181)

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [44]:
# текст запроса
query6_1 = f'''SELECT COUNT(*) AS cnt_vacancies
                FROM vacancies
                WHERE LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%'
'''

In [45]:
# результат запроса
df = pd.read_sql_query(query6_1, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/4288658687.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_1, connection)


cnt_vacancies
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [46]:
# текст запроса
query6_2 = f'''SELECT COUNT(*) AS suitable_vacancies
FROM vacancies
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%') 
    AND 
    (LOWER(name) LIKE '%junior%' OR 
    LOWER(experience) LIKE '%нет опыта%' OR
    LOWER(employment) LIKE '%cтажировка%')
'''

In [47]:
# результат запроса
df = pd.read_sql_query(query6_2, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/155816561.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_2, connection)


suitable_vacancies
0                  51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [48]:
# текст запроса
query6_3 = f'''SELECT COUNT(*) AS suitable_vacancies
FROM vacancies
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%')
    AND
    (LOWER(key_skills) LIKE '%sql%' OR
    LOWER(key_skills) LIKE '%postgres%')
'''

In [49]:
# результат запроса
df = pd.read_sql_query(query6_3, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/2441556487.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_3, connection)


suitable_vacancies
0                 229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [50]:
# текст запроса
query6_4 = f'''SELECT COUNT(*) AS suitable_vacancies
FROM vacancies
WHERE 
    (LOWER(name) LIKE '%data scientist%' OR
    LOWER(name) LIKE '%data science%' OR
    LOWER(name) LIKE '%исследователь данных%' OR
    (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%') OR
    LOWER(name) LIKE '%machine learning%' OR
    LOWER(name) LIKE '%машинн%обучен%')
    AND
    LOWER(key_skills) LIKE '%python%' 
'''

In [51]:
# результат запроса
df = pd.read_sql_query(query6_4, connection)
df

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/1630463327.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_4, connection)


suitable_vacancies
0                 357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [54]:
# текст запроса
query6_5 = f'''
       
SELECT
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_key_skills
    ,COUNT(*) AS total_vacancies
FROM public.vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
   OR LOWER(name) LIKE '%data science%'
   OR LOWER(name) LIKE '%исследователь данных%'
   OR (LOWER(name) LIKE '% ml %' AND LOWER(name) NOT LIKE '%html%')
   OR LOWER(name) LIKE '%machine learning%'
   OR LOWER(name) LIKE '%машинн%обучен%')
   AND key_skills IS NOT NULL
   AND key_skills != '';
'''        


In [55]:
# результат запроса
df = pd.read_sql_query(query6_5, connection)
df 

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/3264641878.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_5, connection)


avg_key_skills  total_vacancies
0            6.41              352

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [58]:
# текст запроса
vacancies=" ((name like '%ML%' and not name like '%HTML%') or (lower (name) like '%data scientist%') or (lower (name) like '%data science%') or (lower (name) like '%исследователь данных%') or (lower (name) like '%machine learning%') or (lower (name) like '%машинн%обучен%')) "
 
query6_6 = f'''
SELECT  experience
        , COUNT(id) AS vacacies_cnt_total
        , SUM(CASE WHEN salary_from IS NOT NULL or salary_to IS NOT NULL then 1 else 0 end ) AS vacacies_with_salary_cnt
        , ROUND(AVG(CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from+salary_to)/2 
                WHEN (salary_from IS NOT NULL and salary_to IS NULL) THEN salary_from
                WHEN (salary_from IS  NULL and salary_to IS NOT NULL) THEN salary_to
                end),0) AS salary
           
FROM vacancies
WHERE  {vacancies}  
GROUP BY experience
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query6_6, connection)
df 

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_48513/269548474.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_6, connection)


experience  vacacies_cnt_total  vacacies_with_salary_cnt    salary
0         Более 6 лет                  34                         0       NaN
1           Нет опыта                  28                         7   74643.0
2  От 1 года до 3 лет                 223                        28  139675.0
3       От 3 до 6 лет                 195                        27  243115.0

***

# выводы по предметному анализу
1771 вакансия имеет отношение к данным( то есть в названии содержатся слова 'data' или 'данн')

51 вакансия подходит для начинающего дата-сайентиста без опыта, ищущего стажировку. 

В 229 вакансиях по  DS в качестве ключевого навыка указан SQL или Postgse
s
B 357 вакансиях по  DS в качестве ключевого навыка указан Pyth
on
В среднем работодатели указывают 6.41 необходимых ключевых навыко
в. 
В среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет может рассчитывать на зп 243115 тыс руб


# дополнительные исследования


как связаны опыт работы, наличие вакансий, среднее количество навыков, зп и тип занятости

In [53]:
query7_2 = f'''
SELECT  experience
        , COUNT(id) AS vacacies_cnt_total
        , ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_key_skills
        , employment as type_employnment
        , ROUND(AVG(CASE WHEN (salary_from IS NOT NULL AND salary_to IS NOT NULL) THEN (salary_from+salary_to)/2 
                WHEN (salary_from IS NOT NULL and salary_to IS NULL) THEN salary_from
                WHEN (salary_from IS  NULL and salary_to IS NOT NULL) THEN salary_to
                end),0) AS salary
FROM public.vacancies
WHERE (LOWER(name) LIKE '%data scientist%'
   OR LOWER(name) LIKE '%data science%'
   OR LOWER(name) LIKE '%исследователь данных%'
   OR (LOWER(name) LIKE '% ml %' AND LOWER(name) NOT LIKE '%html%')
   OR LOWER(name) LIKE '%machine learning%'
   OR LOWER(name) LIKE '%машинн%обучен%')
   AND key_skills IS NOT NULL
   AND key_skills != ''
Group by type_employnment, experience
'''  

In [54]:
# результат запроса
df = pd.read_sql_query(query7_2, connection)
df 

/var/folders/h1/3kts5j755glf5trngng_cjl80000gn/T/ipykernel_75115/3423091397.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query7_2, connection)


experience  vacacies_cnt_total  avg_key_skills  \
0         Более 6 лет                  19            5.42   
1           Нет опыта                  16            7.44   
2  От 1 года до 3 лет                 163            6.87   
3       От 3 до 6 лет                 142            5.90   
4           Нет опыта                   4            5.50   
5           Нет опыта                   4            9.50   
6  От 1 года до 3 лет                   2            4.00   
7       От 3 до 6 лет                   2            5.50   

      type_employnment    salary  
0     Полная занятость       NaN  
1     Полная занятость   82600.0  
2     Полная занятость  155421.0  
3     Полная занятость  254477.0  
4           Стажировка   37000.0  
5  Частичная занятость   72500.0  
6  Частичная занятость       NaN  
7  Частичная занятость  119000.0

# можно сделать следующие выводы:
### для соискателей без опыта:


1. это единственная категория, которую работодатели готовы брать на стажировку, но при этом, если работодатель готов предоставить неопытному специалисту работу на полную занятость, то такой соискатель должен обладать большим количеством компетенций, чем более опытные ( 5.9 против 7.44 ключевых навыка)
2. Разница в вакансиях с частичной занятостью еще более разительная: в среднем 9.5 ключевых навыков ожидают от соискателя без опыта и только 4.0 от соискателя с опытом работы от 1 до 3
3. При этом средняя зарплата на полной или частичной занятости отличается незначительно

## для прочих категорий
1. в отличии от новичков, для соискателей с опытом работы с 1 до 6 требуется в среднем меньше ключевых навыков для частичной занятости чем для полной.
2. для категории более 6 лет предлагаются вакансии только с полной занятостью и зп вероятнее всего обговаривается в ходе собеседования/работы
3. для специалистов с опытом 1-3 года при частичной занятости также зп вероятно обсуждаются на собеседовании
4. Для специалистов с опытом работы 3-6 лет при полной и частичной занятости зп отличается более чем в 2 раза

## Общий вывод по проекту


В целом, результаты анализа показывают активность и потребность в IT-специалистах, особенно в области обработки данных и дата-сайенс
1. .	Анализ показал, что большинство предложений сосредоточено в крупных городах, таких как Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. Что может быть связано как с нахождением самих работодателей в большинстве своем в крупных городах или с боеле высокой  концентрацией там нужных специалист
2. 2.	Менее половины вакансий содержат данные о заработной плате (24073 из 49197), что может указывать на то, что зарплата обсуждается на более поздних стадиях собеседован
3. 3.	 На сайте почти 1771 вакансия для DS и лишь 51 для junior что подтверждает высокую потребность в опытных специали
4. В тах этой области, от соискателя ожидаются высокие компетенции (в среднем подбирается 6.41 необходимых ключевых на
5. 
4.	Меньше всего требуются кандидаты с опытом более 6 лет, при этом наиболее часто востребованы специалисты с 1-3 годами стажа. Это может означать, что работодатели ищут баланса между опытом и стоимостью канддатых навыков)
